Add Import to Operations Pro Locations Page

swalton00 opened the pull request:
The current Operations Pro has an Export Locations which creates a CVS file containing all the Locations and Tracks, with their attributes. This PR adds an Import menu item and the code to support. The import uses a file selected by the user which should be in the format of the file generated by the export. If a Location or Track does not currently exist, it is created and the the attributes are applied to the new Track.

This is a useful tool to migrate the tracks from one Operations Pro profile or system to another, without migrating the entire set of data. An import of the Roster data will create any missing locations, but does not migrate track attribute. An import of the Locations  will migrate the attributes and can be followed by an import of the roster if desired.

Not all attributes are imported. Serveral attributes contain references to other tracks which may not yet exist. These attributes, such as Pool Names and Alternate Track Names are not imported. Restrictions on pickups and setouts, as well as Schedule, Train and Route references are also no imported.

The Test routine simply exports the Location and test whether the file can be successfully imported. There is no mechanism to delete a Location, so it is not a simple matter to create a better test.

danielb987 commented:
Static analysis failed with these errors:
```
1. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 48)
	private static final int FIELD_SHIP_LOAD_OPTION = 12;
	                         ^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_SHIP_LOAD_OPTION is not used
----------
2. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 49)
	private static final int FIELD_SHIPS = 13;
	                         ^^^^^^^^^^^
The value of the field ImportLocations.FIELD_SHIPS is not used
----------
3. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 50)
	private static final int FIELD_SET_OUT_RESTRICTIONS = 14;       // not used
	                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_SET_OUT_RESTRICTIONS is not used
----------
4. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 51)
	private static final int FIELD_RESTRICTIONS_1 = 15;
	                         ^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_RESTRICTIONS_1 is not used
----------
5. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 52)
	private static final int FIELD_PICK_UP_RESTRICTIONS = 16;       // not used
	                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_PICK_UP_RESTRICTIONS is not used
----------
6. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 53)
	private static final int FIELD_RESTRICTIONS_2 = 17;
	                         ^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_RESTRICTIONS_2 is not used
----------
7. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 54)
	private static final int FIELD_SCHEDULE_NAME = 18;
	                         ^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_SCHEDULE_NAME is not used
----------
8. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 55)
	private static final int FIELD_SCHEDULE_MODE = 19;
	                         ^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_SCHEDULE_MODE is not used
----------
9. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 56)
	private static final int FIELD_ALTERNATE_TRACK = 20;
	                         ^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_ALTERNATE_TRACK is not used
----------
10. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 57)
	private static final int FIELD_POOL_NAME = 21;
	                         ^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_POOL_NAME is not used
----------
11. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 61)
	private static final int FIELD_TRACK_DESTINATIONS = 25;
	                         ^^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_TRACK_DESTINATIONS is not used
----------
12. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 62)
	private static final int FIELD_DESTINATIONS = 26;
	                         ^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_DESTINATIONS is not used
----------
13. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 63)
	private static final int FIELD_SWAP_DEFAULT = 27;
	                         ^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_SWAP_DEFAULT is not used
----------
14. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 64)
	private static final int FIELD_EMPTY_DEFAULT_LOADS = 28;
	                         ^^^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_EMPTY_DEFAULT_LOADS is not used
----------
15. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 65)
	private static final int FIELD_EMPTY_CUSTOM_LOADS = 29;
	                         ^^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_EMPTY_CUSTOM_LOADS is not used
----------
16. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 66)
	private static final int FIELD_GENERATE_SPUR = 30;
	                         ^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_GENERATE_SPUR is not used
----------
17. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 67)
	private static final int FIELD_GENERATE_ANY_SPUR = 31;
	                         ^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_GENERATE_ANY_SPUR is not used
----------
18. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 68)
	private static final int FIELD_GENERATE_STAGING = 32;
	                         ^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_GENERATE_STAGING is not used
----------
19. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 69)
	private static final int FIELD_BLOCK_CARS_BY_PICKUP = 33;
	                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
The value of the field ImportLocations.FIELD_BLOCK_CARS_BY_PICKUP is not used
----------
20. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 75)
	public void run() {
	            ^^^^^
The method run() of type ImportLocations should be tagged with @Override since it actually overrides a superclass method
----------
21. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 88)
	int locationsAdded = 0;
	    ^^^^^^^^^^^^^^
The value of the local variable locationsAdded is not used
----------
22. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 205)
	int itemNum = 0;
	    ^^^^^^^
The value of the local variable itemNum is not used
----------
23. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 249)
	String optionValue = "";
	       ^^^^^^^^^^^
The value of the local variable optionValue is not used
```

danielb987 commented on java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java:
```diff
@@ -0,0 +1,343 @@
+    DivisionManager divisionManager = InstanceManager.getDefault(DivisionManager.class);
+
+    private static final int FIELD_LOCATION = 0;
+    private static final int FIELD_TRACK = 1;
+    private static final int FIELD_TYPE = 2;
+    private static final int FIELD_LENGTH = 3;
+    private static final int FIELD_DIVISION = 4;
+    private static final int FIELD_SERVICED_BY = 5;
+    private static final int FIELD_ROLLING_STOCK = 6;
+    private static final int FIELD_ORDER = 7;
+    private static final int FIELD_ROAD_OPTION = 8;
+    private static final int FIELD_ROADS = 9;
+    private static final int FIELD_LOAD_OPTION = 10;
+    private static final int FIELD_LOADS = 11;
+    private static final int FIELD_SHIP_LOAD_OPTION = 12;
+    private static final int FIELD_SHIPS = 13;
+    private static final int FIELD_SET_OUT_RESTRICTIONS = 14;       // not used
+    private static final int FIELD_RESTRICTIONS_1 = 15;
+    private static final int FIELD_PICK_UP_RESTRICTIONS = 16;       // not used
+    private static final int FIELD_RESTRICTIONS_2 = 17;
+    private static final int FIELD_SCHEDULE_NAME = 18;
+    private static final int FIELD_SCHEDULE_MODE = 19;
+    private static final int FIELD_ALTERNATE_TRACK = 20;
+    private static final int FIELD_POOL_NAME = 21;
+    private static final int FIELD_IGNORE_MINIMUM = 22;
+    private static final int FIELD_TRACK_BLOCKING_ORDER = 23;
+    private static final int FIELD_PLANNED_PICK_UPS = 24;
+    private static final int FIELD_TRACK_DESTINATIONS = 25;
+    private static final int FIELD_DESTINATIONS = 26;
+    private static final int FIELD_SWAP_DEFAULT = 27;
+    private static final int FIELD_EMPTY_DEFAULT_LOADS = 28;
+    private static final int FIELD_EMPTY_CUSTOM_LOADS = 29;
+    private static final int FIELD_GENERATE_SPUR = 30;
+    private static final int FIELD_GENERATE_ANY_SPUR = 31;
+    private static final int FIELD_GENERATE_STAGING = 32;
+    private static final int FIELD_BLOCK_CARS_BY_PICKUP = 33;
+    private static final int FIELD_COMMENT = 34;
+    private static final int FIELD_COMMENT_BOTH = 35;
+    private static final int FIELD_COMMENT_PICKUPS = 36;            // not used
+    private static final int FIELD_COMMENT_SETOUTS = 37;            // not used
```
I think it would be good if these are shared with the export tool. Otherwise the export tool might be changed while the import tool remains the same, causing the import failing to import the exported data.

@DanielBoudreau 
What's your view of this? If you think it's good as it is, I'm happy with it.

DanielBoudreau commented:
We can update the export and import locations later to use the same set of offsets.

DanielBoudreau commented:
This isn't true:

The Test routine simply exports the Location and test whether the file can be successfully imported. There is no mechanism to delete a Location, so it is not a simple matter to create a better test.

Use:

LocationManager.deregister(location)

Your test should delete all 5 locations after the export and then confirm their existence after the import.

DanielBoudreau commented:
Thanks for writing this import routine.  I wasn't interested in do it.

DanielBoudreau commented:
You could change the field offset fields from private to protected to get around the static analysis errors.

swalton00 commented:
> This isn't true:
> 
> The Test routine simply exports the Location and test whether the file can be successfully imported. There is no mechanism to delete a Location, so it is not a simple matter to create a better test.
> 
> Use:
> 
> LocationManager.deregister(location)
> 
> Your test should delete all 5 locations after the export and then confirm their existence after the import.

I originally wrote the code to do exactly that,, except that I could never find a way to actually delete the locations (I totally missed "deregister(location)".  I'm testing the code now to do that. I have also changed the private annotation on the static fields to protected. I knew they were not used, but I didn't want to delete them as I concur with the thoughts about sharing them with the Export code.

swalton00 commented:
I didn't want to close the pull request -- I wanted to close the writing of the comment. Is there a way to undo that -- or do I just open a new PR?

DanielBoudreau commented:
Reopened.

swalton00 commented:
I changed the code to use deregister(location) and ran into another issue. The export used JUnitOperationsUtil.loadFiveLocations which does in fact add 5 locations. But Export only exports 3 of them becuase there are only 3 tracks! Two of five locations have no tracks, so they don't get exported. Should I fix the JUnitOperationsUtil to create the tracks -- shouldn't that be a different git change?

DanielBoudreau commented:
Use that info to confirm that only 3 locations are exported / imported.  Don’t fix loadFIveLocations, it could break other test code.

Dan

From: ***@***.***>
Sent: Friday, January 7, 2022 10:25 PM
To: ***@***.***>
Cc: Daniel ***@***.***>; State ***@***.***>
Subject: Re: [JMRI/JMRI] Add Import to Operations Pro Locations Page (PR #10552)


I changed the code to use deregister(location) and ran into another issue. The export used JUnitOperationsUtil.loadFiveLocations which does in fact add 5 locations. But Export only exports 3 of them becuase there are only 3 tracks! Two of five locations have no tracks, so they don't get exported. Should I fix the JUnitOperationsUtil to create the tracks -- shouldn't that be a different git change?

—
Reply to this email directly, view it on GitHub<https://github.com/JMRI/JMRI/pull/10552#issuecomment-1007876843>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADODNJ2PPN4LNQD7P4FITWTUU6VCBANCNFSM5LP2BLWQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

DanielBoudreau commented:
You could use createSevenNormalLocations() instead.  That creates tracks for each location.

Dan

From: ***@***.***>
Sent: Friday, January 7, 2022 10:25 PM
To: ***@***.***>
Cc: Daniel ***@***.***>; State ***@***.***>
Subject: Re: [JMRI/JMRI] Add Import to Operations Pro Locations Page (PR #10552)


I changed the code to use deregister(location) and ran into another issue. The export used JUnitOperationsUtil.loadFiveLocations which does in fact add 5 locations. But Export only exports 3 of them becuase there are only 3 tracks! Two of five locations have no tracks, so they don't get exported. Should I fix the JUnitOperationsUtil to create the tracks -- shouldn't that be a different git change?

—
Reply to this email directly, view it on GitHub<https://github.com/JMRI/JMRI/pull/10552#issuecomment-1007876843>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADODNJ2PPN4LNQD7P4FITWTUU6VCBANCNFSM5LP2BLWQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

danielb987 commented:
Static analysis failed with:
```
1. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 75)
	public void run() {
	            ^^^^^
The method run() of type ImportLocations should be tagged with @Override since it actually overrides a superclass method
----------
2. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 88)
	int locationsAdded = 0;
	    ^^^^^^^^^^^^^^
The value of the local variable locationsAdded is not used
----------
3. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 205)
	int itemNum = 0;
	    ^^^^^^^
The value of the local variable itemNum is not used
----------
4. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 249)
	String optionValue = "";
	       ^^^^^^^^^^^
The value of the local variable optionValue is not used
```

swalton00 commented:
> Static analysis failed with:
> 
> ```
> 1. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 75)
> 	public void run() {
> 	            ^^^^^
> The method run() of type ImportLocations should be tagged with @Override since it actually overrides a superclass method
> ----------
> 2. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 88)
> 	int locationsAdded = 0;
> 	    ^^^^^^^^^^^^^^
> The value of the local variable locationsAdded is not used
> ----------
> 3. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 205)
> 	int itemNum = 0;
> 	    ^^^^^^^
> The value of the local variable itemNum is not used
> ----------
> 4. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrit/operations/locations/tools/ImportLocations.java (at line 249)
> 	String optionValue = "";
> 	       ^^^^^^^^^^^
> The value of the local variable optionValue is not used
> ```

I'll add a new commit with those errors fixed as well as fixing the test to use the creatSevenNormalLocations

swalton00 commented:
I fixed the static analysis issue (I believe I got all the items). I also corrected the test to use the createSevenNormalLocations and it now finds and imports 7 locations. I also caught and fixed a bug while doing so. I have committed those changes and pushed to my repository.

danielb987 commented:
Thanks.

Static analysis fails on:

`Error:  Medium: jmri.jmrit.operations.locations.tools.ImportLocations.run() invokes inefficient new Integer(String) constructor; use Integer.valueOf(String) instead [jmri.jmrit.operations.locations.tools.ImportLocations, jmri.jmrit.operations.locations.tools.ImportLocations, jmri.jmrit.operations.locations.tools.ImportLocations, jmri.jmrit.operations.locations.tools.ImportLocations] At ImportLocations.java:[line 158]Another occurrence at ImportLocations.java:[line 280]Another occurrence at ImportLocations.java:[line 293]Another occurrence at ImportLocations.java:[line 304] DM_NUMBER_CTOR``

Error:  Low: Useless condition: it's known that inputLine.length >= 4 at this point [jmri.jmrit.operations.locations.tools.ImportLocations] At ImportLocations.java:[line 175] UC_USELESS_CONDITION`

swalton00 commented:
I've changed the new Integer to Integer.valueOf(). However I don't like the idea of removing the check for the length of inputLine.

Why were these not pointed out in prior runs. It's bad enough that I missed seeing some of the issues, but to have it keep finding issues is extremely frustrating. And what will happen with the codeclimate issues? I'm not willing to invest that much time doing those changes.

danielb987 commented:
Codeclimate issues is not required to succeed. You see a `Required` label on most but not all the CI tests. And it's only those that are Required that's needed to pass.

Static analysis runs a couple of different tests which is probably why it didn't find the latest issues in the previous runs. If one test fail, static analysis aborts and doesn't run the next test.

bobjacobsen commented:
You're almost there.  Code Climate is advisory, not mandatory.  

For next time, it is possible to run these checks yourself to catch them before you push. 
```
mvn clean test -U -P travis-spotbugs --batch-mode --file=pom.xml
mvn javadoc:javadoc -U --batch-mode --file=pom.xml
```

(The first one runs both SpotBug and Checkstyle)

swalton00 commented:
What about that one static issue — I know it’s going to be true but what if
FIELD_DIVISION is changed in ExportLocations? I don’t like to violate the
pattern of check length before using the value.

On Sat, Jan 8, 2022 at 3:08 PM Daniel ***@***.***> wrote:

> Codeclimate issues is not required to succeed. You see a Required label
> on most but not all the CI tests. And it's only those that are Required
> that's needed to pass.
>
> Static analysis runs a couple of different tests which is probably why it
> didn't find the latest issues in the previous runs. If one test fail,
> static analysis aborts and doesn't run the next test.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/JMRI/JMRI/pull/10552#issuecomment-1008130764>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABO2O3V2HB4KLJ6CIKBCYP3UVCKUDANCNFSM5LP2BLWQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
-- 
Scott Walton

bobjacobsen commented:
> What about that one static issue — I know it’s going to be true but what if FIELD_DIVISION is changed in ExportLocations? I don’t like to violate the pattern of check length before using the value.

There are several approaches:
- The one I prefer, but I understand most people don't, is to add a JUnit test that confirms FIELD_DIVISION still has the right value, i.e. hasn't changed size too much
- Change the check to an `assert`. Those are checked at test-time and will catch the change.
- Suppress the check
```
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
```
Then, on the method, 
```
@SuppressFBWarnings(value = "UC_USELESS_CONDITION", justification = "a runtime check is desired because...")
```

The reason I prefer doing it in tests is that I'd rather catch the problem early and fix it _before_ it hits somebody using the code.

swalton00 commented:
I pushed one last commit. I added a new Test to verify that FIELD_DIVISION is still 4 and commented out the check for the length of the inputLine (and removed the appropriate brace). I ran spotbugs locally and it didn't find anything.
